### PR TITLE
[Feature] Implement `LoRARite` Optimizer

### LINKED ISSUE #498: [Optimizer] LoRA Done RITE: Robust Invariant Transformation Equilibration for LoRA Optimization
https://arxiv.org/abs/2410.20625
https://github.com/gkevinyen5418/LoRA-RITE

### PR DESCRIPTION
**Pull request by mfazrinizar**
## Problem (Why?)

Closes #498.

`pytorch-optimizer` does not yet support LoRA-RITE, the optimizer proposed in "LoRA Done RITE: Robust Invariant Transformation Equilibration for LoRA Optimization".

## Solution (What/How?)

Implemented `LoRARite` as a `BaseOptimizer` optimizer for alternating LoRA factor pairs, matching the reference optimizer's invariant preconditioning flow: basis rotation, unmagnified gradient updates, matrix second-moment tracking, optional escape correction, first-moment tracking, update capping/skipping, coupled weight decay, and optional LoRA factor balancing.

The implementation supports closures, `maximize`, sparse-gradient rejection, complex-parameter rejection, configurable LoRA rank dimensions, loader integration through `load_optimizer('lorarite')`, and `create_optimizer(..., 'lorarite')`.

Added package registration/export, README/docs table entries, API docs, changelog note, package keyword, supported-optimizer count update, a shared training recipe, and focused tests for helper math, state initialization, rich option paths, skipped updates, validation, sparse/complex errors, and loader behavior.

## Other changes (bug fixes, small refactors)

No unrelated refactors or behavior changes.

## Notes

Nothing so far.

## Checklist

- [x] Make sure to run `make format` before commit
- [x] My code adheres to the style guidelines of this project (`make check` shows no errors)
- [x] Both new and existing unit tests pass successfully on my local environment by running `make test`
- [x] I have made the necessary changes to the documentation

### CODE REVIEW

File: pytorch_optimizer/optimizer/lora_rite.py [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+            self.init_group(group)
+            group['step'] += 1
+
+            for param_left, param_right in self.iter_lora_pairs(group):
+                if param_left.grad is None or param_right.grad is None:
+                    continue
+
+                pair_info = self.build_pair_info(group, param_left, param_right)
+                update_left, update_right = pair_info[1], pair_info[2]
+                grad_norm_sq = grad_norm_sq.to(update_left.device).add_(torch.linalg.norm(update_left).pow(2))
```
Comment (kozistr) [inline pytorch_optimizer/optimizer/lora_rite.py]: how about creating `grad_norm_sq` tensor on the same device as `update_left` at initialization (line 438)? fetching it to another device every time could be costly

File: pytorch_optimizer/optimizer/lora_rite.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+            ('apply_escape', apply_escape),
+            ('maybe_inf_to_nan', maybe_inf_to_nan),
+            ('balance_param', balance_param),
+            ('maximize', maximize),
+        ):
+            if not isinstance(value, bool):
+                raise ValueError(f'{name} must be a boolean')
+
+        for name, value in (('lora_l_dim', lora_l_dim), ('lora_r_dim', lora_r_dim)):
+            if not isinstance(value, int) or isinstance(value, bool):
```
Comment (kozistr) [inline pytorch_optimizer/optimizer/lora_rite.py]: i guess we don't need boolean check!

```suggestion
            if not isinstance(value, int):
```

File: pytorch_optimizer/optimizer/lora_rite.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+
+        for name, value in (
+            ('relative_epsilon', relative_epsilon),
+            ('apply_escape', apply_escape),
+            ('maybe_inf_to_nan', maybe_inf_to_nan),
+            ('balance_param', balance_param),
+            ('maximize', maximize),
+        ):
+            if not isinstance(value, bool):
+                raise ValueError(f'{name} must be a boolean')
```
Comment (kozistr) [inline pytorch_optimizer/optimizer/lora_rite.py]: we don't need extra runtime type checks here for these options since we already have type hints. of course, type-hint is not a strict checker, but we can just skip additional checks here.

File: pytorch_optimizer/optimizer/lora_rite.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+
+            for param_left, param_right in self.iter_lora_pairs(group):
+                if param_left.grad is None or param_right.grad is None:
+                    continue
+
+                pair_info = self.build_pair_info(group, param_left, param_right)
+                update_left, update_right = pair_info[1], pair_info[2]
+                grad_norm_sq = grad_norm_sq.to(update_left.device).add_(torch.linalg.norm(update_left).pow(2))
+                grad_norm_sq = grad_norm_sq.add_(torch.linalg.norm(update_right).pow(2))
+                pair_infos.append((group, param_left, param_right, pair_info))
```
Comment (kozistr) [inline pytorch_optimizer/optimizer/lora_rite.py]: maybe we can utilize the `state` to save `param_left` and `param_right` instead of saving all info into `pair_infos`

**Review by kozistr — COMMENTED** [review]
thanks for your contributions :) leave some comments!

File: pytorch_optimizer/optimizer/lora_rite.py (hunk 1 above)
Comment (mfazrinizar) [reply pytorch_optimizer/optimizer/lora_rite.py]: Makes sense. I'll update `grad_norm_sq` so it is lazily initialized with `update_left.new_zeros(())`, so we avoid moving the accumulator each pair.

**Comment by kozistr** [comment]
looks good to me. could you please resolve the conflicts here too?

**Review by kozistr — APPROVED** [review]
thank you for your contribution!
